### DISCUSSION TITLE
Wrap function that can convert .ipynb <> my format into NotebookSerializer

### DISCUSSION THREAD

**Discussion by arogozhnikov** [microsoft/vscode-jupyter#10174] [Questions and Answers]
My previous related question: https://github.com/microsoft/vscode-jupyter/discussions/10121

My team uses a custom format for Jupyter notebooks which is convertible to and from .ipynb.
I 'd like to have the same functionality for vscode. 

To write a custom NotebookSerializer I should implement all the conversion utilities from here: https://github.com/microsoft/vscode-jupyter/blob/4339372e6dcc518719f3e6132a28b8baae296d13/src/notebooks/helpers.ts

But I only want to add one step my format <> ipynb, and I'd like vs code to deal with ipynb <> vs code representation. 
How can I achieve that?

**Comment by DonJayamanne** [comment]
I believe you are after something similar to jupytext, where the backing file is a Python file or the like, but you'd like it to be converted to an ipynb and use the notebook experience for execution and the like (whilst the file is still saved as a plain text)

Is that correct? If thats correct, then unfortunately we don't support such functionality.

**Reply by arogozhnikov** [reply]
That's correct, I use such format for jupyter and would like to have support for vs code too.

> unfortunately we don't support such functionality.

uh, ok. Do you have any recommendations how can I implement this myself?

**Comment by arogozhnikov** [comment]
any changes here? I'm coming back to this again and again.

How it looks for me now:
- ipynb is a standard, more or less
- vscode has its own representation, with conversion <> ipynb

There are serializers that can work with ipynb, but they can't be reused to have vs code <> ipynb <> custom format.

Instead one has to copy a good chunk of vs code to get this working. 

@DonJayamanne how about this idea:

allow extensions to not provide their custom serializer, but specify to use default ipynb serializer, and provide converter functions in typescript that will convert 'ipynb bytes <> custom format bytes' and back? 

That sounds like +several of lines in ipynb serializer to handle this, and should unblock jupytext and, well, others

**Reply by DonJayamanne** [reply]
@arogozhnikov 
Thanks but I'm failing to understand your requirement here.
Are you after the following:
* You have some custom file format
* You open that and want it to be displayed exactly like a Jupyter notebook (ipynb) file in VS Code
* You want all of the exact same functionality provided today for Jupuyter notebooks in VS Code (i.e. allow users to execute code in notebooks, etc), 
* However at the end of the day the underlying file is NOT ipynb, but is a different format, is that correct?

**Reply by arogozhnikov** [reply]
Answers: yes, yes, yes, and yes. 

And there is an available function for conversion ipynb <> custom format

**Reply by DonJayamanne** [reply]
Leave this with me, I'll look into this soon.

**Reply by DonJayamanne** [reply]
I'll re-visit https://github.com/microsoft/vscode-jupyter/issues/1237
Please update the above issue so we have the discussion in one place

**Reply by arogozhnikov** [reply]
@DonJayamanne I've tried to implement this with current API of vscode by creating a custom NotebookSerializer - but only to figure out that it demands a custom kernel too! What is supposed to be 'just add two functions bytes -> bytes' turned into a huge project. There should be some reasonable way

**Reply by DonJayamanne** [reply]
> hat is supposed to be 'just add two functions bytes -> bytes' turned into a huge project. There should be some reasonable way

Sorry you ran into this, but I do not think I lead you down this path, I.e. i do not believe I said it was just two fucntions bytes -> bytes.
As mentioned earlier, this isn't supported today, and the original issue was #1237 (which happens to be closed in favour of https://github.com/microsoft/vscode-jupyter/issues/1240).

See here https://github.com/microsoft/vscode-jupyter/discussions/10174#discussioncomment-3942650

**Reply by arogozhnikov** [reply]
I've read all these threads several times, they aren't the same request, there is no hint to solution. This discussion is from 2022, and issues are from 2020.

Well, ok, that issue is not of interest to you - now I'm trying to make use of APIs. But these APIs don't support reasonably scoped extension - and I don't see why. Problem for me is literally to insert two converting functions + custom extension, but I don't see a way to achieve this.

**Reply by DonJayamanne** [reply]
> Well, ok, that issue is not of interest to you - now I'm trying to make use of APIs. 

I'm sorry you feel this way. But I'm not sure where you got this. I'm engaging with you, and trying to help you., 
in this case, trying to help you understand that the issue will need to be addressed in #1240, and a simple 3rd party extension will not work.


> But these APIs don't support reasonably scoped extension - and I don't see why. Problem for me is literally to insert two converting functions + custom extension, but I don

This isn't possible with a simple extension.
Notebooks work off a serializer, the Jupyter notebook serializer will always read files from disc and create a Jupyter notebook.
There's no way to get notebooks to read from two different sources. 
This requires a lot more work and analysis, hence the issue #1240

This is one such extension i built as an exploration and found a number of limitations, https://marketplace.visualstudio.com/items?itemName=donjayamanne.vscode-jupytext
Since then others have built other extensions similar to this, but still have the same limitations.

**Comment by DonJayamanne** [comment]
Closing this as a duplicate of https://github.com/microsoft/vscode-jupyter/issues/1240
